app/lib/features/chat_ng/widgets/events/text_message_event.dart

```diff
@@ -1,153 +1,112 @@
-import 'package:acter/common/actions/open_link.dart';
-import 'package:acter/common/themes/acter_theme.dart';
 import 'package:acter/common/themes/app_theme.dart';
-import 'package:acter/features/chat/widgets/pill_builder.dart';
-import 'package:acter/l10n/generated/l10n.dart';
+import 'package:acter/common/toolkit/html/render_html.dart';
+import 'package:acter/features/chat_ng/utils.dart';
 import 'package:acter_flutter_sdk/acter_flutter_sdk_ffi.dart' show MsgContent;
 import 'package:flutter/material.dart';
-import 'package:flutter_matrix_html/flutter_html.dart';
-import 'package:flutter_matrix_html/text_parser.dart';
 import 'package:flutter_riverpod/flutter_riverpod.dart';
-import 'package:markdown/markdown.dart' as md;
 
-enum TextMessageType { regular, reply, emoji, notice }
-
-class TextMessageEvent extends ConsumerWidget {
+class TextMessageEvent extends ConsumerStatefulWidget {
   final String roomId;
   final MsgContent content;
-  final TextMessageType _type;
-  final bool _isMe; // Only needed for emoji messages to determine style
+  final bool isNotice;
+  final bool isReply;
   final Widget? repliedTo;
 
-  const TextMessageEvent.inner({
+  const TextMessageEvent({
     super.key,
     required this.content,
     required this.roomId,
-    required TextMessageType type,
-    bool isMe = false,
+    this.isNotice = false,
+    this.isReply = false,
     this.repliedTo,
-  }) : _type = type,
-       _isMe = isMe;
+  });
 
-  factory TextMessageEvent.emoji({
-    Key? key,
-    required MsgContent content,
-    required String roomId,
-    required bool isMe,
-  }) {
-    return TextMessageEvent.inner(
-      key: key,
-      content: content,
-      roomId: roomId,
-      type: TextMessageType.emoji,
-      isMe: isMe,
-    );
-  }
+  @override
+  ConsumerState<TextMessageEvent> createState() => _TextMessageEventState();
+}
 
-  // Factory constructor for reply messages
-  factory TextMessageEvent.reply({
-    Key? key,
-    required MsgContent content,
-    required String roomId,
-  }) {
-    return TextMessageEvent.inner(
-      key: key,
-      content: content,
-      roomId: roomId,
-      type: TextMessageType.reply,
-    );
-  }
+class _TextMessageEventState extends ConsumerState<TextMessageEvent> {
+  late String body;
+  String? bodyFormatted;
```

Comment:
stateful to keep the info around and not parse the HTML more often than necessary. Additionally we don't have to read the content twice to check for emojis and then render it, but just put into memory once.